Edits to lower_layer.md (#2343)
mirage/irmin: doc/irmin-pack/design/lower_layer.md

1. Modify GC API and process to allow unlimited-history stores to move history into the lower layer instead of discarding it
1. Modify data reading process to use lower layer when data is not found in the upper layer

These changes and related ones are discussed in the following sections.

### New Terms

The following new terms are introduced:

- **Upper layer**: the area of a store that contains recent history and receives new commits: the "recent" area.
- **Lower layer**: the area of an unlimited-history store that holds history discarded by GC from the upper layer: the "archive" area.
- **Volume**: a sub-area of the **lower layer** for storing a contiguous portion of a store's history. Multiple volumes can be created to reduce loss of data locality.
    - **Appendable volume**: the read-write volume in the **lower layer**

## New Store Configuration Values

Currenlty, an `irmin-pack` store uses a configuration value named `root` to know the path on disk to use for reading and writing its files. To support a lower layer, the configuration is extended by adding a value named `lower_root`. This is a path on disk for reading and writing volumes.

## Lower Layer

> The lower layer is the archive area for an unlimited history store. It stores the commit history that the GC process discared from the upper layer.

Reviewer: Small historic typo, it should be "discarded"

> 

Author: Li 74  :: Could this be: 
"Since it contains a contiguous commit history, there are no commit gaps between its first and last commit" 
?

1. Create empty `store.1.suffix` in upper layer
1. Create empty prefix and mapping in upper layer
1. Update upper layer control file

>     - `status` is GC'd

Reviewer 2: ```suggestion
    - `status` is `Gced`
```
Sorry this specific `Gced` actually refers to a term in the code so we can't change it :P

Author: Oh! Good to know!

This scenario is detected by comparing the generation number (the `gen` part of the name) of the volume's control file and the generation in the upper control file. If they are the same, this is resolved by swapping the appendable volume's control file on restart via a file rename. If they are not, it is unlinked.

This scenario highlights a motivating reason for `volume.gen.control`. It gives us a way to coordinate both the volume's control files and upper layer and to recover in the scenarios where one or the other fails to update.

**Moment C**
![](./imgs/lower_layer/bXuEjOi.png)

This moment is after the volume's control file has been renamed but before files have been unlinked (in the upper layer).

Nothing changes from current behaviour. Files are unlinked on node restart.

### Validations

The following validations are performed as a part of the new GC process:

- The appendable volume's `end_offset` should be less than or equal to the upper layer's `suffix_start_offset` before appending data to the volume. This should never happen in pracitce, but this validation protects against writing offsets that have already been written to the volume.

### GC API Usage

> The new APIs are intended to make using a lower layer with multiple volumes as straightforward as possible. Since every store will start with an empty volume in its lower layer, a client only needs to run a GC as in previous versions of `irmin-pack` by calling `Gc.run`. Data will be moved to the lower layer's volume automatically. 

Author: There might be a missing word here...

"The new APIs are intended to make ______ using a lower layer..."
What does it intend to make?

Reviewer: I understand it as direct a translation from french: The APIs intend to make the use of a lower layer that has multiple volumes as easy as possible.

Author: Oh now I understand as I reread it again. Thanks.

- How do we determine which volume contains a commit hash?
- Once we have loaded a commit, how do we continue reading its data (nodes and contents) from its volume?

### Finding a Commit

**Upper Layer**

> The starting point for finding a commit in `irmin-pack` is to lookup its key in the index. Assuming the store has a record in its commit hash's index, the key will provide a global offset for finding the commit. Lets call this simply `offset`.

Reviewer 2: ```suggestion
The starting point for finding a commit in `irmin-pack` is to lookup its key in the index. Assuming the store has a record of the commit hash in its index, the key will provide a global offset for finding the commit. Lets call this simply `offset`.
```

I'm not sure what's the proper English way to formulate this sentence, so I tried switching words around... but don't hesitate if you have a better idea! (but the "of" is not a possessive)

Author: Let's just put it back the way it was, then. 🙂

> When the appendable volume's control file is renamed from `volume.gen.control` to `volume.control`, readers see the new end bounds of a volume at the last readable offset of `end_offset`, and the mapping file is extended to contain all entries to the new `mapping_end_poff`. Until this happens, readers read the data as it was before GC.

Author: ```suggestion
When the appendable volume's control file is renamed from `volume.gen.control` to `volume.control`, readers see a volume's new end bounds at `end_offset`'s last readable offset, and the mapping file is extended to contain all the new `mapping_end_poff` entries. Until this happens, readers read the data as it was before GC.
```

Is this still accurate?
